src/emulator/flash.c

```diff
@@ -0,0 +1,217 @@
+                        if (!fn_80061BC0(pFLASH->pStore, buffer, i, 0x80)) {
+                            return false;
+                        }
+                    }
+                } else if (!fn_80061BC0(pFLASH->pStore, buffer, pFLASH->unk_14, 0x80)) {
+                    return false;
+                }
+            }
+
+            pFLASH->unk_18 = 0;
+            break;
+        case 0x0:
+            break;
+        case 0xE1000000:
+            pFLASH->unk_18 = 1;
+            pFLASH->flashStatus = 0x11118001;
+            break;
+        case 0x3C000000:
+            pFLASH->unk_14 = -1;
+            break;
+        case 0x4B000000:
+            pFLASH->unk_14 = (*pData << 7) & 0x7FFFFF80;
+            break;
+        case 0x78000000:
+            pFLASH->flashStatus = 0x11118008;
+            pFLASH->unk_18 = 4;
+            break;
+        case 0xB4000000:
+            pFLASH->unk_18 = 3;
+            break;
+        case 0xA5000000:
+            pFLASH->unk_18 = 3;
+            pFLASH->unk_14 = (*pData << 7) & 0x7FFFFF80;
+            pFLASH->flashStatus = 0x11118004;
+            break;
+        case 0xF0000000:
+            pFLASH->flashStatus = 0x11118004;
+            pFLASH->unk_18 = 2;
+            break;
+        default:
+            return false;
+    }
+
+    pFLASH->flashCommand = *pData;
+    return true;
+}
+
+static bool flashPut64(Flash* pFLASH, u32 nAddress, s64* pData) { return true; }
+
+static bool flashGet8(Flash* pFLASH, u32 nAddress, s8* pData) { return true; }
+
+static bool flashGet16(Flash* pFLASH, u32 nAddress, s16* pData) { return true; }
+
+static bool flashGet32(Flash* pFLASH, u32 nAddress, s32* pData) {
+    if (nAddress == 0x08000000) {
+        *pData = pFLASH->flashStatus;
+    } else {
+        *pData = (nAddress & 0xFFFF) | ((nAddress << 16) & ~0xFFFF);
+    }
+
+    return true;
+}
+
+static bool flashGet64(Flash* pFLASH, u32 nAddress, s64* pData) { return true; }
+
+static bool flashGetBlock(Flash* pFLASH, CpuBlock* pBlock) {
+    void* pRAM;
+
+    if ((pBlock->nAddress0 & 0xFF000000) == 0x08000000) {
+        if (!ramGetBuffer(SYSTEM_RAM(gpSystem), &pRAM, pBlock->nAddress1, &pBlock->nSize)) {
+            return false;
+        }
+
+        if (pFLASH->unk_18 == 1) {
+            *((u32*)pRAM + 0) = pFLASH->flashStatus;
+            *((u32*)pRAM + 1) = 0xC2001D;
+        } else if (pFLASH->unk_18 == 2) {
+            fn_80061B88(pFLASH->pStore, pRAM, pBlock->nAddress0 - 0x08000000, pBlock->nSize);
+        }
+    } else if ((pBlock->nAddress1 & 0xFF000000) == 0x08000000) {
+        pFLASH->nOffsetRAM = pBlock->nAddress0;
+    }
+
+    if (pBlock->pfUnknown != NULL && !pBlock->pfUnknown(pBlock, 1)) {
+        return false;
+    }
+
+    return true;
+}
+
+bool fn_80045260(Flash* pFLASH, s32 arg1, void* arg2) {
+    //! TODO: fake match?
+    if (pFLASH == NULL) {
+        return false;
+    }
+
+    return !!fn_80061B88(pFLASH->pStore, arg2, (arg1 * 8) & 0x7F8, 8);
```

Comment:
Do you have a scratch for this? I forget, is this compiler smart enough to turn
```
if (foo()) {
  return true;
}
return false;
```
(or the reverse) into this? If so I think we should prefer that over `!!` whenever possible

Reply:
I think this is actually ``return foo() ? true : false``, I noticed it produced the same behavior than ``!!`` while matching SDK stuff yesterday

Reply:
change to
```
if (!foo()) {
  return false;
}
return true;
```
instead? I think it's a bit easier to read, and based on oot-gc it's probably what it was originally.
(also remove the "fake match?" comment? seems fairly real to me with that)